### ISSUE TITLE
Installed binaries can be used only in the mode they were configured

### ISSUE THREAD

**Issue by denesb** [scylladb/seastar#627]
Binaries installed via `ninja install` or `make install` can only be used in the mode they were configured in. For example release binaries can only be used in applications that are configured in release mode. Using a binary compiled in other modes will possibly lead to segmentation fault or other obscure failures.

**Comment by denesb** [comment]
The underlying issue is that due to the different `#ifdef`s in seastar's code, depending on the compile-type flags a different binary will be generated. This is a well-known problem in the C and C++ world.
This results in the application consuming a seastar installation being constrained to building the configuration the seastar installation was built with.
I'm not sure how we can solve this, if there is even a good way. At the very least we should document this in our tutorials on how a seastar installation can be consumed.

**Comment by denesb** [comment]
For an example of the failures can occur see: https://github.com/scylladb/seastar/issues/599
Note however that this failure happened while using the library via `pkgconfig` and was caused by user mistake, still it is the same kind of crash users experiencing this issue will face.

**Comment by denesb** [comment]
Ideally it should be possible to co-install multiple configurations, `cmake` picking the binaries for the build-mode the application is built with. Also, it should refuse to configure if the requested mode is not present on the machine. I think this should be possible to do, no?

**Comment by hakuch** [comment]
If I understand correctly, the header files themselves won't change across build modes.

Therefore, I think we can solve this problem by using a different output name for the library depending on the build mode.

**Comment by hakuch** [comment]
It's possible to do something like this:

```
set_target_properties (seastar
  PROPERTIES
    DEV_POSTFIX "-dev"
    DEBUG_POSTFIX "-debug")
```

such that installing Seastar in "dev" mode installs `libseastar-dev.a` and installing in "debug" mode installs `libseastar-debug.a`.

For example, this is part of the content of `SeastarTargets.cmake` after configuring in "dev" mode:

```
# Import target "Seastar::seastar" for configuration "Dev"
set_property(TARGET Seastar::seastar APPEND PROPERTY IMPORTED_CONFIGURATIONS DEV)
set_target_properties(Seastar::seastar PROPERTIES
  IMPORTED_LINK_INTERFACE_LANGUAGES_DEV "CXX"
  IMPORTED_LOCATION_DEV "/home/jhaberku/src/seastar/build/dev/libseastar-dev.a"
  )
```

and this is part of the generated `seastar.pc`:

```
seastar_libs=${libdir}/libseastar-dev.a 
```


However, installing two different modes to the same directory results in this message from CMake:

```
-- Old export file "/home/jhaberku/src/seastar/foobar/lib64/cmake/Seastar/SeastarTargets.cmake" will be replaced.  Removing files [/home/jhaberku/src/seastar/foobar/lib64/cmake/Seastar/SeastarTargets-debug.cmake].
```

Therefore, I think with this change we can differentiate  installed Seastar artifacts in the "wrong" mode, but we cannot support installing multiple modes to the same directory. That is, they would need to be installed in different locations.

**Comment by hakuch** [comment]
What do you think, @denesb?

**Comment by senior7515** [comment]
@hakuch i guess the SeastarConfig.cmake would have to find the -debug and the -release versions too, for  cmake downstream libs, or how do you suggest one would go about that.

**Comment by hakuch** [comment]
@senior7515: `SeastarConfig.cmake` references `SeastarTargets.cmake`, and that file (which is generated automatically by CMake) references the mode-specific target file. Nothing needs to change.

**Comment by senior7515** [comment]
nice.

**Comment by nyh** [comment]
Here is my two cents on how this thing is *supposed* to work in the pkg-config world (I can't say anything about cmake).

The idea is that the library *and* seastar.pc are installed at the same time, and fit each other. If the installed library is in debug mode and requires applications to compile with -DDEBUG, or whatever, the seastar.pc will say that. If the installed library is in release mode and requires other build options - seastar.pc will say *that*.

If we provide some way of installing two different versions of the library in parallel, e.g., in two different directories or two different names, then we also need to install two versions of seastar.pc. We already have a different seastar.pc for each build mode (I insisted on this to allow in-place use of the built library without installation). If there are errors in those seastar.pc files, we should fix them.

**Comment by denesb** [comment]
@hakuch I think this is a good start and should be already useful on its own to avoid mistakes. I think we should still try to find a way to co-install multiple binaries, but that is a nice-to-have.

**Comment by denesb** [comment]
@nyh the problem is that there is nothing stopping the user from setting arbitrary preprocessor flags and end up in a situation where the code compiled into their application (via the headers) is incompatible from what is in `libseastar.a`. We should do our best to prevent this... if at all possible. I'm not sure what is the accepted practice w.r.t. this problem in the community. Surely other libraries have faced this. I know e.g. that boost ships several different binaries that amount to different build configurations but I don't know how they do it, or what others are doing.

**Comment by nyh** [comment]
Part of the solution is not to rely on generic-sounding preprocessor macros (e.g., "DEBUG") that the application author might accidentally use on his own, and call our flags SEASTAR_DEBUG or similar, and actually we're already doing this. A user who redefines a flag called "SEASTAR_DEBUG" manually is just shooting his own foot, and we don't need to "solve this". The seastar.pc for each mode will set the SEASTAR_DEBUG flag differently automatically, and the user should never set this flag manually. He just gets the correct flags automatically from `pkg-config --cflags .../seastar.pc`.

I don't know what you are referring to Boost. In most libraries (not specifically boost), the "debug" version of the library just means that it has debugging symbols, and timetimes - more assertions. Most libraries do not have different ABIs for the debugging version, so they are not worried by what worries you here.
